File: so/os/os.go
// Package os provides a platform-independent interface to operating system
// functionality. The design is Unix-like, although the error handling is Go-like;
// failing calls return values of type error rather than error numbers.
package os

import (
	"solod.dev/so/c"
	"solod.dev/so/errors"
	"solod.dev/so/time"
)

// IO-related errors that can be returned by functions in this package.
var ErrClosed = errors.New("os: file already closed")
var ErrExist = errors.New("os: file already exists")
var ErrIsDir = errors.New("os: is a directory")
var ErrNotDir = errors.New("os: not a directory")
var ErrNotExist = errors.New("os: no such file or directory")
var ErrPermission = errors.New("os: permission denied")

// ErrIO is a generic I/O error that is returned when the error
// does not match any of the other, more specific errors.
var ErrIO = errors.New("os: i/o error")

// Chdir changes the current working directory to the named directory.
func Chdir(dir string) error {
	if chdir(dir) != 0 {
		return mapError()
	}
	return nil
}

// Chmod changes the mode of the named file to mode.
// If the file is a symbolic link, it changes the mode of the link's target.
func Chmod(name string, mode FileMode) error {
	pmode := makePosixMode(mode)
	if chmod(name, pmode) != 0 {
		return mapError()
	}
	return nil
}

// Chown changes the numeric uid and gid of the named file.
// If the file is a symbolic link, it changes the uid and gid of the link's target.
// A uid or gid of -1 means to not change that value.
func Chown(name string, uid, gid int) error {
	if chown(name, uid_t(uid), gid_t(gid)) != 0 {
		return mapError()
	}
	return nil
}

// Chtimes changes the access and modification times of the named
// file, similar to the Unix utime() or utimes() functions.
// A zero [time.Time] value will leave the corresponding file time unchanged.
func Chtimes(name string, atime time.Time, mtime time.Time) error {
	var asec, ansec, msec, mnsec int64
	if atime.IsZero() {
		ansec = utimeOmit
	} else {
		asec, ansec = atime.Unix(), int64(atime.Nanosecond())
	}
	if mtime.IsZero() {
		mnsec = utimeOmit
	} else {
		msec, mnsec = mtime.Unix(), int64(mtime.Nanosecond())
	}
	if os_utimens(name, asec, ansec, msec, mnsec) != 0 {
		return mapError()
	}
	return nil
}

// Lchown changes the numeric uid and gid of the named file.
// If the file is a symbolic link, it changes the uid and gid of the link itself.
func Lchown(name string, uid, gid int) error {
	if lchown(name, uid_t(uid), gid_t(gid)) != 0 {
		return mapError()
	}
	return nil
}

// Link creates newname as a hard link to the oldname file.
func Link(oldname, newname string) error {
	if link(oldname, newname) != 0 {
		return mapError()
	}
	return nil
}

// Mkdir creates a new directory with the specified name and permission
// bits (before umask).
func Mkdir(name string, perm FileMode) error {
	pmode := makePosixMode(perm)
	if mkdir(name, pmode) != 0 {
		return mapError()
	}
	return nil
}

// Readlink returns the destination of the named symbolic link.
// If the link destination is relative, Readlink returns the relative path
// without resolving it to an absolute one.
//
// Writes the result into buf. The returned string is a view into buf.
func Readlink(buf []byte, name string) (string, error) {
	n := readlink(name, c.CharPtr(&buf[0]), uintptr(len(buf)))
	if n < 0 {
		return "", mapError()
	}
	return string(c.Bytes(&buf[0], n)), nil
}

// Remove removes the named file or (empty) directory.
func Remove(name string) error {
	if remove(name) != 0 {
		return mapError()
	}
	return nil
}

// Rename renames (moves) oldpath to newpath. If newpath already exists
// and is not a directory, Rename replaces it. If newpath already exists
// and is a directory, Rename returns an error. OS-specific restrictions
// may apply when oldpath and newpath are in different directories.
// Even within the same directory, on non-Unix platforms Rename
// is not an atomic operation.
func Rename(oldpath, newpath string) error {
	if rename(oldpath, newpath) != 0 {
		return mapError()
	}
	return nil
}

// SameFile reports whether fi1 and fi2 describe the same file.
// For example, on Unix this means that the device and inode fields
// of the two underlying structures are identical.
func SameFile(fi1, fi2 FileInfo) bool {
	return fi1.dev == fi2.dev && fi1.ino == fi2.ino
}

// Symlink creates newname as a symbolic link to oldname.
func Symlink(oldname, newname string) error {
	if symlink(oldname, newname) != 0 {
		return mapError()
	}
	return nil
}

// Truncate changes the size of the named file.
// If the file is a symbolic link, it changes the size of the link's target.
func Truncate(name string, size int64) error {
	if truncate(name, size) != 0 {
		return mapError()
	}
	return nil
}

// mapError maps errno to a sentinel error.
func mapError() error {
	if errno == eACCES {
		return ErrPermission
	}
	if errno == eEXIST {
		return ErrExist
	}
	if errno == eISDIR {
		return ErrIsDir
	}
	if errno == eNOENT {
		return ErrNotExist
	}
	if errno == eNOTDIR {
		return ErrNotDir
	}
	if errno == ePERM {
		return ErrPermission
	}
	return ErrIO
}

File: so/os/os.c
//go:build ignore
#include <dirent.h>
#include <errno.h>
#include <fcntl.h>
#include <string.h>
#include <sys/stat.h>
#include "so/builtin/builtin.h"

// Stat result - flat struct filled by C helpers.
typedef struct {
    int64_t size;
    uint32_t mode;
    int64_t modSec;
    int64_t modNsec;
    uint64_t dev;
    uint64_t ino;
    bool ok;
} os_statResult;

// os_lstat fills result from lstat().
static os_statResult os_lstat(const char* path) {
    struct stat st;
    if (lstat(path, &st) != 0) return (os_statResult){.ok = false};
    return (os_statResult){
        .size = st.st_size,
        .mode = st.st_mode,
        .modSec = st.st_mtime,
        .modNsec = 0,  // fields differ on macos and linux, set to 0 for now
        .dev = st.st_dev,
        .ino = st.st_ino,
        .ok = true,
    };
}

// os_stat fills result from stat().
static os_statResult os_stat(const char* path) {
    struct stat st;
    if (stat(path, &st) != 0) return (os_statResult){.ok = false};
    return (os_statResult){
        .size = st.st_size,
        .mode = st.st_mode,
        .modSec = st.st_mtime,
        .modNsec = 0,  // fields differ on macos and linux, set to 0 for now
        .dev = st.st_dev,
        .ino = st.st_ino,
        .ok = true,
    };
}

// os_utimens sets access and modification times using utimensat.
// A tv_nsec of UTIME_OMIT leaves the corresponding time unchanged.
static int os_utimens(const char* path, int64_t asec, int64_t ansec, int64_t msec, int64_t mnsec) {
    struct timespec times[2] = {
        {.tv_sec = asec, .tv_nsec = ansec},
        {.tv_sec = msec, .tv_nsec = mnsec},
    };
    return utimensat(AT_FDCWD, path, times, 0);
}

// readdir result - one entry at a time.
typedef struct {
    int32_t nameLen;
    uint8_t dtype;
    bool ok;
} os_readdirResult;

// os_readdir_next reads the next directory entry.
// Copies d_name into buf. Returns {nameLen, dtype, ok}.
static os_readdirResult os_readdir_next(DIR* dir, char* buf, size_t bufsize) {
    errno = 0;
    struct dirent* ent = readdir(dir);
    if (ent == NULL) return (os_readdirResult){.ok = false};
    size_t n = strlen(ent->d_name);
    if (n >= bufsize) n = bufsize - 1;
    memcpy(buf, ent->d_name, n);
    buf[n] = '\0';
    return (os_readdirResult){.nameLen = (int32_t)n, .dtype = ent->d_type, .ok = true};
}
